File: 367 Valid Perfect Square.py

# 367 Valid Perfect Square:
# https://leetcode.com/problems/valid-perfect-square/
print("# Problem Link: https://leetcode.com/problems/valid-perfect-square/")

class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        start_index = 1
        end_index = num

        while start_index <= end_index:
            mid_index = (start_index + end_index) // 2
            mid_squared = mid_index * mid_index

            if num == mid_squared:
                return True
            elif num > mid_squared:
                start_index = mid_index + 1
            elif num < mid_squared:
                end_index = mid_index - 1

        return False    

sol = Solution()
print(sol.isPerfectSquare(16))  # True
print(sol.isPerfectSquare(17))  # False

File: 367 Valid Perfect Square.swift
// 367 Valid Perfect Square
// https://leetcode.com/problems/valid-perfect-square/

class Solution {
    func isPerfectSquare(_ num: Int) -> Bool {
        var start_index = 1
        var end_index = num
        
        while start_index <= end_index {
            // Define a mid and mid squared index for binary search
            let mid_index = (start_index + end_index) / 2
            let mid_squared = mid_index * mid_index
            
            if mid_squared == num {
                return true
            } else if mid_squared < num {
                start_index = mid_index + 1
            } else {
                end_index = mid_index - 1
            }
        }
        return false
    }
}

let sol = Solution()
print(sol.isPerfectSquare(16))
print(sol.isPerfectSquare(17))

